### ISSUE TITLE
Redo swiss matching code

### ISSUE THREAD

**Issue by longtimegone** [PennyDreadfulMTG/gatherling_fork#19]
The existing code is a kludge that falls apart if the event is too small, it really should be something else, maybe a stable roommates algorithm.

**Comment by stash86** [comment]
If it's possible, I want the code doesn't directly posting the nexr round's pairing, but instead the code have to check if there are 2 or more byes in that pairing. If yes (2 or more byes exists), the code then redo the pairing again, until it reaches 0 or 1 bye only or reaches the maximum tries allowed (let's say you set it to 3 tries maximum). Whichever the condition that stop the iteration, the code will post the latest pairing for that next round.

**Comment by bakert** [comment]
There's some certainly-not-optimized code that only goes up to (I think) 11 players for swiss pairings that I wrote at https://github.com/bakert/tournament/blob/master/classes/pairings.php

It does not give two players the bye under any circumstances (!)

It may or may not be a useful starting point for this.

It would require perf optimizations (or rewriting) to work for larger tournaments as it can take up to 30s to pair an 11-man pod if there have been draws. (Without draws it's very fast. Basically if there's an "optimal" set of pairings it will find it fast. If there isn't it calculates a score for every possible set of pairings before lumping for one of the best suboptimal pairings. There are probably better algorithms.)

**Comment by silasary** [comment]
https://twitter.com/CubeApril/status/945857357845794818
«My take on byes: early round byes should tiebreak at the average match-win-percentage (MWP) of the people who lost in those rounds, instead of being ignored. 

The fact that these early byes give you both the fewest chances to lose *and* give you the best breakers is pretty lame.»

**Comment by stash86** [comment]
I have seen and experience cases where 2-1 with 1 bye finished above a fully-played 2-1, and it really irked some players where you lose prizes because of luck factor

**Comment by stash86** [comment]
This is certainly more important on swiss-only PREs, like vintage, legacy and modern. And in some cases, affect top 8 cut for other PREs

**Comment by bakert** [comment]
Deviating from the practice used in Magic Online and paper tournaments seems like a mistake to me. If people don't like the results you can at least point to an authority that does it that way. And people will complain about any system.

**Comment by silasary** [comment]
This should also allow for manual seeds, and/or pulling seeds from the season leaderboard.

**Comment by stash86** [comment]
Can this be done ASAP? I have encountered a weirdpairing today

Swiss standing looked like this

Rank | Player | Match Points | OMW % | PGW % | OGW % | Matches Played | Byes
-- | -- | -- | -- | -- | -- | -- | --
1 | NewHJ | 9 | 0.554 | 1.000 | 0.381 | 3 | 0
2 | morpphling | 9 | 0.500 | 0.800 | 0.455 | 2 | 1
3 | IronicGentleman | 6 | 0.667 | 0.625 | 0.618 | 3 | 0
4 | Yokai_ | 6 | 0.667 | 0.571 | 0.648 | 3 | 0
5 | AEFabricio | 6 | 0.554 | 0.571 | 0.512 | 3 | 0
6 | tomsimons | 6 | 0.443 | 0.500 | 0.450 | 3 | 0
7 | Socanelas | 6 | 0.442 | 0.625 | 0.292 | 3 | 0
8 | RedderTory (drop) | 3 | 0.833 | 0.000 | 0.813 | 2 | 1
9 | br_laern | 3 | 0.667 | 0.286 | 0.600 | 3 | 0
10 | kalcin66rocks | 3 | 0.554 | 0.444 | 0.468 | 3 | 0
11 | maCHOOga | 3 | 0.443 | 0.429 | 0.387 | 3 | 0
12 | stash86 | 0 | 0.556 | 0.250 | 0.542 | 3 | 0
13 | DuckX (drop) | 0 | 0.500 | 0.333 | 0.472 | 2 | 0

And the pairing?
morpphling vs. NewHJ
Yokai_ vs. Socanelas
tomsimons vs. IronicGentleman
AEFabricio vs. kalcin66rocks
maCHOOga has the BYE
br_laern vs. stash86

How a 3-pointer got the bye over a 0-pointer?

**Comment by stash86** [comment]
A 6 rounds of swiss mock event based on @bakert 's MaxWeight algorhytm
https://helhalmshop.tk/mock_swiss_round.php

Is it look okay?

Edit : ignore it for now. I saw a higher point player get a bye over 0-point player. Have to re-work the code a little bit it seems

**Comment by stash86** [comment]
Okay, I think I get this works now. Please help me try to break this pairing code
https://helhalmshop.tk/mock_swiss_round.php?players=51&rounds=5

Please do change the number of players and rounds. Thanks

**Comment by bakert** [comment]
If you put in a really excessive number of rounds (I tried 900) then it
seems one player will repeatedly get the bye (much more than the other
players). While that's not much of a practical concern it may reveal a
problem in a more realistic scenario.

**Comment by stash86** [comment]
@bakert fixed. So for now, once all players has received at least one bye, the system will give bye to a random player with the lowest points for that round. So it's possible for a player to receive multiple byes in a row if their point gap with the next lowest point is 6 or more.

**Comment by bakert** [comment]
Did you update the test page with the latest code?

Strictly speaking in
https://helhalmshop.tk/mock_swiss_round.php?players=3&rounds=7 Player1
should get the BYE in round 5.

**Comment by bakert** [comment]
And should Player2 get the bye in round 6? Not sure.

https://helhalmshop.tk/mock_swiss_round.php?players=3&rounds=7

**Comment by stash86** [comment]
This is what I get using that same setting
7 rounds of swiss. The left players always won
There is 3 players 
Pairing for Round 1 
Player3 (0) v Player1 (0) 
Player2 (0) BYE 
Bye receivers are Player2, 

There is 3 players 
Pairing for Round 2 
Player2 (3) v Player3 (3) 
Player1 (0) BYE 
Bye receivers are Player2, Player1, 

There is 3 players 
Pairing for Round 3 
Player2 (6) v Player1 (3) 
Player3 (3) BYE 
Bye receivers are Player2, Player1, Player3, 

There is 3 players 
Pairing for Round 4 
Player3 (6) v Player2 (9) 
Player1 (3) BYE 
Bye receivers are Player2, Player1, Player3, Player1, 

There is 3 players 
Pairing for Round 5 
Player2 (9) v Player3 (9) 
Player1 (6) BYE 
Bye receivers are Player2, Player1, Player3, Player1, Player1, 

There is 3 players 
Pairing for Round 6 
Player1 (9) v Player3 (9) 
Player2 (12) BYE 
Bye receivers are Player2, Player1, Player3, Player1, Player1, Player2, 

There is 3 players 
Pairing for Round 7 
Player3 (9) BYE 
Player2 (15) v Player1 (12) 
Bye receivers are Player2, Player1, Player3, Player1, Player1, Player2, Player3, 

Round 5 is correct, round 6 not so much.

Basically once all players has received a bye, the code are less stable. Idk if I want to do more hack to prevent this case when this case (all active players has received bye) have a 0.0001% of happening.

And we have no guideline on how to handle bye in that situation. Do a player allowed to receive multiple byes, or do the bye record is resetted to 0?

**Comment by bakert** [comment]
My concern is only that it may reveal structural weaknesses in our
approach. I’m not actually concerned with what happens in a 7 round
tournament with 3 people :)

Best of all would be to set up some tests based on “bad” scenarios from
previous gatherling tournaments. we also ought to have tests involving
draws as some formats allow that.

**Comment by stash86** [comment]
draws are allowed only on the last round of swiss, so it's not a concern of mine, although I will do some testing based on past events once I can integrate the code into gatherling system. It gonna take some time to modify the code to suit gatjerling system

On 27 Aug 2018, at 20.50, Thomas David Baker <notifications@github.com> wrote:

My concern is only that it may reveal structural weaknesses in our
approach. I’m not actually concerned with what happens in a 7 round
tournament with 3 people :)

Best of all would be to set up some tests based on “bad” scenarios from
previous gatherling tournaments. we also ought to have tests involving
draws as some formats allow that.

**Comment by silasary** [comment]
Historically, there have been events that allow for draws whenever.

We need to be able to support people making Round 3 (or earlier) draws.

**Comment by stash86** [comment]
Okay, draws are added to the testing code. So far the pairing behaves normally. Please let me know if you encounters any error

**Comment by bakert** [comment]
I'd like to see more randomization in the results as the patterns you have
chosen may not reveal any bugs. Or (better) some test cases from real life
stupid gatherling cases which should be pretty easy to source.

But looks good.

**Comment by stash86** [comment]
Done integrating the new pairing code into gatherling. For an event with 10 active players, it took ~9s to pair a round. Idk if all the individual DB calls were the cause for the slow time. Gonna play a little bit more to test it.

**Comment by bakert** [comment]
If it’s possible I think it would be wise to killswitch the change so we
can go back to the old way without a release

**Comment by stash86** [comment]
Do you mean I keep both pairing code and let the admin (ltg and silasary) to be able to choose which code to be used at any time?

**Comment by silasary** [comment]
Note that if it takes 9 seconds, you're going to have to:
* Make sure the pairing happens in a way that the user isn't stuck on a white page for the duration
* Do something about https://github.com/PennyDreadfulMTG/gatherling/issues/25, which will become *a lot* worse if the margin for overlap becomes that long.

**Comment by silasary** [comment]
And yeah, a toggle would very much be appreciated.

**Comment by bakert** [comment]
yeah some way to switch that doesn't require a deploy that katelyn (or any
host preferable) would be able to switch.

i don't think we should release if a 10 person tournament on local takes
9s. let's figure out why and fix.

**Comment by stash86** [comment]
Hmm.. Using the old code, I get the same ~9s time to pair that same round. Is it my computer lagging that cause the long processing time?

**Comment by bakert** [comment]
Brute forcing pairings is pretty expensive. My fairly naïve code in my
tournament-running software for up to pods of 11 can take 30s to pair a
round 2 where there's a single draw in r1 (it really wants to pair those
players again but it is not allowed to). I never bothered to make it smart
because beyond 11 players it makes multiple pods. The super naïve version
is O(n!) which gets really slow really quickly. I would expect the blossom
algorithm stuff to be faster as I think that's O(n^2). Is the time taken
inside the blossom code or outside? Do you hit the database more than
once/inside a loop?

**Comment by stash86** [comment]
So for the same procedure of "Delete matches and re-pair round" pressed for 5x for each of pairing code on my local comp
old code : 6.71, 6.32, 7.27, 6.28, 6.85 for an average of 6.686s
blossome pairing : 6.36, 7.03, 6.66, 5.72, 6.09 for an average of 6.372s

so in theory, the blossom pairing works faster by a small margin than the old pairing. I don't know what causing both of the codes to take ~6-7s for pairing, but I guess I should test in on a live gatherling on my server

**Comment by stash86** [comment]
The time I used up there are the "Finish" time from Developer Console>Network on Chrome

**Comment by stash86** [comment]
Okay, so you can try the new pairing system at http://helhalmshop.tk/gatherling/. I'll send the login info by PM. Few things

1. In the Event Setting's structure drop-down, I add "Swiss (Blossom)" as another event type in addition to "Swiss". "Swiss" would use current gatherling pairing code

2. I would need you to add "Swiss (Blossom)" to the ENUM of type column in the subevents table

3. I did a simple test of clicking on the "Delete matches and re-pair round" and get the time for the new page to fully loaded (the time based on Developer Tools timer in Chrome)
Blossom : 3.21s, 3.13, 2.87, 3.12, 2.63 -> average of 2.992s
Old code: 2.63, 4.22, 3.12, 3.01, 2.58 -> average of 3.112s

So there is no significant increase in pairing time (it even shows slight decrease of loading time), which is good. Please do try it out yourself.

**Comment by stash86** [comment]
So I made a [test](https://github.com/stash86/gatherling/blob/stash-eventtests/tests/EventsRealCases001Test.php) to simulate [this event](http://www.gatherling.com/event.php?name=Penny%20Dreadful%20Saturdays%209.08&view=match) . I set each pairing code to re-pair the 4th round 200 times each and check whether the pairing has BYE pairing or not. And the result are

`Out of 200 tries, there are 0 cases of double byes using Blossom Pairing. Out of 200 tries, there are 67 cases of double byes using Brute Pairing`

Gonna find another past event to be tested

**Comment by stash86** [comment]
So I done a second test based on [this event](http://www.gatherling.com/eventreport.php?event=Classic%20Heirloom%2012.66) where a player received their 2nd bye on round 3. And the result wasn't good

Out of 200 tries, there are 144 cases of a player receiving more than 1 bye using Blossom Pairing. Out of 200 tries, there are 200 cases of a player receiving more than 1 bye using Brute Pairing

**Comment by stash86** [comment]
A little tweak and we are back in business 🔥 

Out of 200 tries, there are 0 cases of a player receiving more than 1 bye using Blossom Pairing. Out of 200 tries, there are 200 cases of a player receiving more than 1 bye using Brute Pairing.

Any idea what other situations that I need to test the code against? @bakert @silasary

**Comment by bakert** [comment]
As the known-bad state is more than one person getting a bye you could write code that generates tournaments at random and let it run for a long time (a day?) It could emit something when it finds a tournament where more than one player got a bye in a round.

**Comment by stash86** [comment]
Test number 3 based on [this event](http://www.gatherling.com/eventreport.php?event=Classic%20Heirloom%2012.74) where we have 2 players received a bye on round 3.

Out of 200 tries, there are 0 cases of more than one player receive a bye in a round and 0 cases of a player receiving more than 1 bye using Blossom Pairing. Out of 200 tries, there are 134 cases of more than one player receive a bye in a round and 0 cases of a player receiving more than 1 bye using Brute Pairing.

**Comment by stash86** [comment]
4th test based on [this event](http://www.gatherling.com/eventreport.php?event=Classic%20Heirloom%2012.49) where a player received 2 byes.

Out of 200 tries, there are 0 cases of more than one player receive a bye in a round and 0 cases of a player receiving more than 1 bye using Blossom Pairing. Out of 200 tries, there are 0 cases of more than one player receive a bye in a round and 200 cases of a player receiving more than 1 bye using Brute Pairing.

**Comment by stash86** [comment]
So I think I have finalized the blossom pairing code,

I have set up a live php test [here](https://helhalmshop.tk/gatherling/tests/EventsRealCasesTest.php) and the code for the test page can be found [here](https://github.com/stash86/gatherling/blob/blossompairing/tests/EventsRealCasesTest.php). It would take around 5-6 mins for it to finish all 4 test cases (it needs to do 1600 pairings.

**Comment by bakert** [comment]
Nice :)

**Comment by bakert** [comment]
@longtimegone trying every possible avenue to get ahold of you. We need to add the RNA cards to Gatherling as a matter of urgency. Can you get in touch? bakert@gmail.com
